src/messaging/ExchangeMgr.cpp

```diff
@@ -144,13 +151,15 @@ void ExchangeManager::OnReceiveError(CHIP_ERROR error, const Transport::PeerAddr
 ExchangeContext * ExchangeManager::AllocContext(uint16_t ExchangeId, uint64_t PeerNodeId, bool Initiator,
                                                 ExchangeDelegate * delegate)
 {
+    ExchangeContext * ec = mContextPool;
+
     CHIP_FAULT_INJECT(FaultInjection::kFault_AllocExchangeContext, return nullptr);
 
-    for (auto & ec : ContextPool)
+    for (int i = 0; i < CHIP_CONFIG_MAX_EXCHANGE_CONTEXTS; i++, ec++)
```

Comment:
was the previous implementation not working?

Reply:
Now, we need to share ContextPool to other classes, previous implementation use std::array, which is more difficult to passing this struct over function argument and validate it.
std::array<ExchangeContext, CHIP_CONFIG_MAX_EXCHANGE_CONTEXTS> ContextPool;

Reply:
I don't understand this reasoning.

1) If the issue is that the `std::array` type is just too long to type out, create a typedef for it.
2) If there are other issues with using `std::array`, what are they, exactly?  If we want to share the pool with the reliable message manager, give it a reference to the same array, fine.
3) Even if you do switch to using a C-style array, you can still use a ranged for with it.

Reply:
My original try is to define a reference within ReliableMessageManager 
std::array<ExchangeContext, CHIP_CONFIG_MAX_EXCHANGE_CONTEXTS> & mContextPool;

But I am not able to initialize the reference value within init function. Another workaround is to initialize the reference value within constructor instead of init function. But I am not sure if this binding is always valid, will std relocate element of std::array in some cases like std::vector?

Reply:
> Another workaround is to initialize the reference value within constructor instead of init function. 

This isn't a "workaround"; it's the right thing to do if we want one to alias the other.

> But I am not sure if this binding is always valid,

Why would it be invalid?  It's a reference to the object, so as long as the object survives it's valid.  The caller of the constructor would of course need to ensure it survives.

> will std relocate element of std::array in some cases like std::vector?

Generally no, depending on what you mean by "relocate element".

Reply:
@bzbarsky-apple 

Do you know how to make mContextPool reference initialized before following placeholder function instantiation? 

    /* Placeholder function to run a function for all exchanges */
    template <typename Function>
    void ExecuteForAllContext(Function function)
    {
        for (auto & ec : mContextPool)
        {
            function(ec.GetReliableMessageContext());
        }
    }

I hit the following compile error if I switch to reference method.

    ../../src/messaging/ReliableMessageManager.h:115:26: error: implicit instantiation of undefined template 'std::__2::array<chip::Messaging::ExchangeContext, 8>'
        for (auto & ec : mContextPool)
                         ^

Reply:
Hard to say without seeing the actual code you have there.  Please post it somewhere?

Reply:
`#include <array>` in `ReliableMessageManager.h`

Reply:
Excellent, I am always confused the static instantiation order within the class.